ion-c-sys/src/writer.rs

```diff
@@ -428,97 +461,134 @@ impl Drop for IonCWriterHandle<'_> {
 ///
-/// Specifically, writing fields with Ion C with `ion_writer_write_field`
+/// Specifically, writing annotations and fields with Ion C with
+/// `ion_writer_write_annotations`/`ion_writer_add_annotation` and `ion_writer_write_field`
 /// is particularly problematic because the lifetime of the `ION_STRING`
 /// passed into that function, must be valid until a corresponding
 /// `ion_writer_write_*` function is called for writing a value.
 ///
-/// This context tracks the lifetime of the string slice that a user wants to
-/// write as a field name along with a mutable borrow of the writer handle
-/// and ensures that calls to `ion_writer_write_field` happen before invoking the
-/// `ion_writer_write_*` call for the value in a lifetime correct way.
+/// This context tracks the lifetime of the string slices that a user wants to
+/// write as annotations and the field name along with a mutable borrow of the writer handle
+/// and ensures that calls to `ion_writer_add_annotation`/`ion_writer_write_field`
+/// happen before invoking the `ion_writer_write_*` call for the value in a lifetime correct way.
 ///
 /// Note that this context can be thought of as a partial function for writing
-/// `struct` fields, so doing nothing with it is simply a no-op, and performing
+/// annotations and/or `struct` fields, so doing nothing with it is simply a no-op, and performing
 /// multiple [`IonCValueWriter`](./trait.IonCValueWriter.html) trait method invocations
 /// is the same as writing the field before invoking the value writing methods
 /// (so duplicate fields would be generated in that case).
-pub struct IonCFieldWriterContext<'a, 'b, 'c> {
+pub struct IonCAnnotationsFieldWriterContext<'a, 'b, 'c> {
     handle: &'b mut IonCWriterHandle<'a>,
-    field: &'c str,
+    annotations: Option<&'c [&'c str]>,
+    field: Option<&'c str>,
 }
 
-impl<'a, 'b, 'c> IonCFieldWriterContext<'a, 'b, 'c> {
-    fn new(handle: &'b mut IonCWriterHandle<'a>, field: &'c str) -> Self {
-        Self { handle, field }
+impl<'a, 'b, 'c> IonCAnnotationsFieldWriterContext<'a, 'b, 'c> {
+    fn new_field(handle: &'b mut IonCWriterHandle<'a>, field: &'c str) -> Self {
+        Self {
+            handle,
+            annotations: None,
+            field: Some(field),
+        }
+    }
+
+    fn new_annotations(handle: &'b mut IonCWriterHandle<'a>, annotations: &'c [&'c str]) -> Self {
+        Self {
+            handle,
+            annotations: Some(annotations),
+            field: None,
+        }
+    }
+
+    /// Sets the annotations for the context.
+    pub fn annotations(&mut self, annotations: &'c [&'c str]) -> &mut Self {
+        self.annotations = Some(annotations);
+        self
+    }
+
+    /// Sets the field name for the context.
+    pub fn field(&mut self, field: &'c str) -> &mut Self {
+        self.field = Some(field);
+        self
     }
 }
 
-macro_rules! write_field {
+macro_rules! write_annotations_field {
```

Comment:
```suggestion
macro_rules! write_field_annotations {
```

Reply:
I refactored this into an inline non-public function called `write_annotations_and_field()`, is there a particular reason why you want the ordering of the names here?  I ask because this context is for either field writing, annotation writing, or both.